core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -580,23 +581,29 @@ public PlanOptimizers(
                                 new PruneTableScanColumns(metadata),
                                 new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))));
 
+        Set<Rule<?>> pushIntoTableScanRules = ImmutableSet.<Rule<?>>builder()
+                .addAll(columnPruningRules)
+                .addAll(projectionPushdownRules)
+                .add(new PushProjectionIntoTableScan(metadata, typeAnalyzer))
+                .add(new RemoveRedundantIdentityProjections())
+                .add(new PushLimitIntoTableScan(metadata))
+                .add(new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))
+                .add(new PushSampleIntoTableScan(metadata))
+                .add(new PushJoinIntoTableScan(metadata))
+                .add(new PushAggregationIntoTableScan(metadata))
+                .add(new PushDistinctLimitIntoTableScan(metadata))
+                .add(new PushTopNIntoTableScan(metadata))
+                .build();
         IterativeOptimizer pushIntoTableScanOptimizer = new IterativeOptimizer(
                 ruleStats,
                 statsCalculator,
                 estimatedExchangesCostCalculator,
-                ImmutableSet.<Rule<?>>builder()
-                        .addAll(columnPruningRules)
-                        .addAll(projectionPushdownRules)
-                        .add(new PushProjectionIntoTableScan(metadata, typeAnalyzer))
-                        .add(new RemoveRedundantIdentityProjections())
-                        .add(new PushLimitIntoTableScan(metadata))
-                        .add(new PushPredicateIntoTableScan(metadata, typeOperators, typeAnalyzer))
-                        .add(new PushSampleIntoTableScan(metadata))
-                        .add(new PushJoinIntoTableScan(metadata))
-                        .add(new PushAggregationIntoTableScan(metadata))
-                        .add(new PushDistinctLimitIntoTableScan(metadata))
-                        .add(new PushTopNIntoTableScan(metadata))
-                        .build());
+                pushIntoTableScanRules.stream()
+                        // PushJoinIntoTableScan must run after ReorderJoins to not undo join reordering
+                        // and after DetermineJoinDistributionType because DetermineJoinDistributionType requires
+                        // table statistics which are lost after join pushdown
+                        .filter(rule -> !(rule instanceof PushJoinIntoTableScan))
```

Comment:
nit: instead of an instanceof check, create `pushIntoTableScanRulesExceptJoins` that doesn't contain a Join pushdown, and then, in the other place, `ImmutableSet.builder().addAll(...).add(pushJoinIntoTS)...`

the benefits are

- no instanceof check
- we want to run pushJoinIntoTS in very specifically chosen place (after RJ and DetermineJoinDT and before AddExchanges, which pretty much is very well defined), while all the other rules can be run more than once (they are once _today_)

Reply:
I kept it this way because it looks like the rule declaration order matters (we try them in order) if multiple rules match the PlanNode being optimized.

I can indeed simplify it if I misunderstood the IterativeOptimizer.

Reply:
The ordering of rules in single IterativeOptimizer should not matter. If it matters I would treat that as bug.

Reply:
Simplified.